### ISSUE TITLE
Allow HPA to scale to 0

### ISSUE THREAD

**Issue by shosti** [kubernetes/kubernetes#69687]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

I am trying to set up a HorizontalPodAutoscaler for a worker deployment that scales up and down based on a queue size (which feeds into a custom metric). The queue is quite bursty, so I would like the deployment to be scaled to 0 unless there are messages in the queue. I tried setting `minReplicas` to 0, but got errors like:

```
HorizontalPodAutoscaler.autoscaling "my-autoscaler" is invalid: spec.minReplicas: Invalid value: 0: must be greater than 0
```
(Using the `autoscaling/v2beta1` API). 

**What you expected to happen**:

I would expect it to be allowed for autoscalers to scale deployments down to 0 for these kinds of bursty workflows. Apologies if this has already been requested/discussed--I looked around but only found a vague mention of this in https://github.com/kubernetes/community/blob/master/contributors/design-proposals/autoscaling/horizontal-pod-autoscaler.md and https://github.com/kubernetes/kubernetes/pull/3247.

**How to reproduce it (as minimally and precisely as possible)**:

Try to make a `HorizontalPodAutoscaler` with API `autoscaling/v2beta1` and `minReplicas: 0`.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"archive", BuildDate:"1970-01-01T00:00:01Z", GoVersion:"go1.10.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.7-gke.2", GitCommit:"8d9503f982872112eb283f78cefc6944af640427", GitTreeState:"clean", BuildDate:"2018-09-13T22:19:55Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: GKE

**Comment by ericjee** [comment]
Hi, any updates here?

**Comment by miks** [comment]
Same here. We have quite sporadic background jobs with custom video pipeline. There is no sense to have even 1 pod with high CPU reservation (otherwise cluster node autoscale will not kick in) when the jobs queue is empty.

**Comment by marekbiskup** [comment]
PTAL at Knative serving:
https://github.com/knative/docs/tree/master/serving
It should solve your problems if you don't mind it's version 0.2.0
https://github.com/knative/build/releases

**Comment by akitekt** [comment]
Knative serving can scale to 0, but it requires Istio and Serving components.
If HPA can scale pod to 0, I would choose the simple and easy route for sure.

**Comment by DXist** [comment]
+1

My use case - set of workers that use GPU. I want to run them on limited number of GPUs.

**Comment by Rajat-0** [comment]
I would like to work on this if Sig leads agree to implement this.

**Comment by gogiraghu1** [comment]
Hi,
  Any update on this bug? Looks like this feature for HPA is in high demand until knative completes HPA integration. Please keep us posted.

**Comment by govindKAG** [comment]
this cannot come any sooner

**Comment by Rajat-0** [comment]
https://docs.google.com/document/d/1p_Xlk8e5V32WOBVeRbJqbxhEKcJy_mDu9JgVp5ZNDxs/

I have Prepared above Design Doc for this issue and discussed it in SIG Autoscaler weekly meeting, and as the discussion the Sig owner will discuss the possibility and complexity of the issue with the Networking team.

As per my proposed design:

For scaling up from pod 0 firstly we have to make a change in service controller to gracefully handle the request when no pod is available. Service controller syncs and updates service status and load balancer status and its associated hosts. The traffic to Pods is directed to backend pods via load balancer, and its depends on how that works. In service controller there is no provision to check, existence of Pods.


1.To scale up form 0 we need monitor the request to the Service.
2.First request to the Service will trigger a CreatePod event. 
3.Create buffer until the kube-proxy receives endpoint information for requested service.
4.Controller resolves the Service to some replication controller.
5.The replication controller manager schedules a new pod 
6.The endpoints controller determines that service has a new endpoint and updates Service's endpoints.
7.The kube-proxy receives a watch event with the new endpoint information and updates its routing table.
8.The kube-proxy services the request to the new endpoint.

**Comment by Rajat-0** [comment]
@DXist  I am also working on this issue and I have discussed the above proposal in Sig weekly meeting, and based on how design goes, we both can coordinate on implementation.

**Comment by mikkeloscar** [comment]
> The traffic to Pods is directed to backend pods via load balancer, and its depends on how that works. In service controller there is no provision to check, existence of Pods.

Why does this matter for the general case? You can autoscale pods without a service, and you can currently scale any deployment to 0 (without HPA) and it would have the same "problem".

IMO solve the general problem first and deal with the edge cases as a separate issue. There are many applications which doesn't handle ingress traffic but could benefit from scaling up from 0.

**Comment by DXist** [comment]
@mikkeloscar I am working on general case at the moment to cover work with queue.

@Rajat-0 You might better to coordinate with knative guys and their work to support HPA in service Activator.

**Comment by Rajat-0** [comment]
@DXist thanks for letting me know about Knative's plan to support HPA in service Activator, but I think It would we great if we can have this functionality in K8s/HPA because, as per my my knowledge Knative requires istio and knative solution works for Knative workload.

**Comment by josephburnett** [comment]
FYI: This Knative issues tracks HPA scale-to-zero: https://github.com/knative/serving/issues/3064  There is nothing technical preventing its implementation--just time and energy!

An alternative would be Osiris from Deislabs: https://github.com/deislabs/osiris  This provides general hpa scale-to-zero by injecting a proxy and running an Activator (similar to Knative but without Istio dependency).

**Comment by josephburnett** [comment]
@DXist if I understand your proposal correctly, you would buffer incoming requests in the kube-proxy while the first pod is being created.  Is that right?

**Comment by DXist** [comment]
@josephburnett , request buffering is out of scope of https://github.com/kubernetes/kubernetes/pull/74526
My project uses Rabbitmq. Queue broker acts as a buffer for worker tasks.

For HTTP requests buffering in kube-proxy might be the way to go. Another approach is to introduce external load balancer with request buffering support in similar way as `https://github.com/gptankit/serviceq` does.

**Comment by josephburnett** [comment]
@DXist thanks for the clarification.

I meant to ask the question to @Rajat-0 regarding the outlined plan: https://github.com/kubernetes/kubernetes/issues/69687#issuecomment-467082733

> 3.Create buffer until the kube-proxy receives endpoint information for requested service.

**Comment by wolkjesmaker** [comment]
I could also benefit to scale down to 0. 
No need to have anything running unless there's some custom metric that enabled HPA.

**Comment by Rajat-0** [comment]
@josephburnett Yes, the main idea is to trigger the First pod creation when First request hits, and buffer requests in the kube-proxy while the first pod is being created.

**Comment by josephburnett** [comment]
@Rajat-0 in Knative we've been thinking about how we would integrate with your proposal.  And how to evolve our scale-to-zero system in general.  One change we are making is to separate metrics collection from autoscaling and provide our custom metric (request concurrency) from a custom metrics server.  @markusthoemmes is working on that [here](https://docs.google.com/document/d/1Mvf1aLBbXtqmlYRZkSdS7Mdq0DUTVlvdA9sBlRNFDWs/edit#).

It might be worth getting together to discuss.  To see what the commonalities are.  Any interest in presenting your proposal at the Knative Scaling Working Group?  Wednesday at 9:30am PST (5:30pm CET).

**Comment by josephburnett** [comment]
@Rajat-0 another interesting change happening in Knative networking around scale-to-zero is handling route changes entirely through K8s services ([design](https://docs.google.com/document/d/1byHQL6kePDZq6Qt4MqB-g4BDfiaJsfDPGb5oJEK1Ldo/edit))

**Comment by Rajat-0** [comment]
@josephburnett I'll attend next Knative Scaling Working Group meeting and we can discuss about it there.

**Comment by josephburnett** [comment]
The Knative docs are usually shared with https://groups.google.com/forum/#!forum/knative-dev.  It would be easiest if you joined that group.

**Comment by josephburnett** [comment]
Thanks @Rajat-0 for presenting at the Knative scaling working group!  Notes and recording from the meeting are [here](https://docs.google.com/document/d/1FoLJqbDJM8_tw7CON-CJZsO2mlF8Ia1cWzCjWX8HDAI/edit#heading=h.1itre3nkszl) (join [knative-users](https://groups.google.com/forum/#!forum/knative-users) for read access).

The approach @Rajat-0 outlined would use iptables to route requests for idle services back to the kube-proxy, which would 1) hold the request until service pods are available and 2) emit a metric which the autoscaler would see and use to scale up from zero.

## Prior Art

This is essentially the same approach as taken by Knative (see [this context diagram](https://github.com/knative/serving/blob/de40f6efc63a07a15b2373a80eec7e5552628377/docs/scaling/DEVELOPMENT.md#context)).  Except that Knative runs the proxy as a separate service, the Activator.

Last year @rbitia present [Osiris](https://github.com/deislabs/osiris) scale-to-zero at the Knative scaling working group.  Notes and recording [here](https://docs.google.com/document/d/1FoLJqbDJM8_tw7CON-CJZsO2mlF8Ia1cWzCjWX8HDAI/edit#heading=h.d9wghrqr1m1g).  Osiris takes the same approach for catching requests--running an Activator service that holds requests, triggers scale-up and then proxies them.

Differences are in how scale-up and scale-down are orchestrated.  Knative currently uses Istio routes and will soon have a CRD that manages service endpoints manually ([design doc](https://docs.google.com/document/d/1byHQL6kePDZq6Qt4MqB-g4BDfiaJsfDPGb5oJEK1Ldo/edit#heading=h.wwf0syhtntjh)).  Osiris does the same by "hijacking" the service and manually managing the endpoints ([how it works](https://github.com/deislabs/osiris/blob/6b693280cc799ed60f066249f9cd939555763675/README.md#how-it-works)).

Both Knative and Osiris need to inject a sidecar into the user's Pod in order to observe traffic and scale down when idle.

## Next Steps

I recommend running the Activator-analogue as a separate service / daemonset to avoid complexity in the kube-proxy.  Possibilities include 1) continuing development on Osiris, 2) generalizing the Knative Activator, or 3) building something new.  I am inclined towards using a preexisting code base.  There are many subtle edge cases and considerations in building a scale-to-zero system, so putting more eggs in fewer baskets would be better.

One such edge case is how to trigger immediate scale up on an idle service.  If the autoscaler is periodically scraping metrics to scale up, artificial latency will be introduced by the polling loop.  Knative currently runs a service on the Autoscaler to which the Activator pushes metrics immediately on the first request for an idle service.  But we will be separating decision making from metrics, putting the latter behind a custom metrics provider ([design doc](https://docs.google.com/document/d/1Mvf1aLBbXtqmlYRZkSdS7Mdq0DUTVlvdA9sBlRNFDWs/edit#heading=h.iix47t97buat)).  In this architecture, we would like 1) a watch API for custom metrics and 2) the ability to register multiple custom metrics providers in the cluster.

Another such consideration is how to determine which deployment to scale up when an idle service receives traffic.  In Knative we inject a header at ingress (Istio) which specifies the name of the deployment (there is only one deployment per Knative Revision).  General Kubernetes services can be more various.

**Comment by reyou** [comment]
This use case seems like very close problem what https://kubeless.io/ is resolving.
Scaling from 0 Pod to N Pods.

I would really would like to have this feature in K8S natively, but meanwhile people who are seeking immediate solution may have take a look into this one.

Disclaimer: I haven't tried it yet.

**Comment by palade** [comment]
> 
> 
> This use case seems like very close problem what https://kubeless.io/ is resolving.
> Scaling from 0 Pod to N Pods.
> 
> I would really like to have this feature in K8S natively, but meanwhile people who are seeking immediate solution may have take a look into this one.
> 
> Disclaimer: I haven't tried it yet.

Did you further investigate this issue? I believe the scale to 0 in kubeless is still marked as future work

**Comment by reyou** [comment]
@palade that's correct. The link I referred is a 3rd party unofficial workaround to this feature request.

**Comment by palade** [comment]
@reyou I don't think Kubeless implements this feature yet. Please correct if this is not accurate

**Comment by reyou** [comment]
@palade hmm, what you said seems accurate. Honestly, I watched their showcase video, and I was under impression that scales to zero. So function supposed to get evicted after a time out.

In this ticket,
https://github.com/kubeless/kubeless/issues/982

Following one referred as a alternate 
https://github.com/deislabs/osiris

**Comment by palade** [comment]
@reyou Thank you

**Comment by krelst** [comment]
Just leaving my use case over here: a deployment of which the pods pull messages from gcloud pub/sub, autoscales on number of outstanding messages on the subscription, and should scale back to 0 pods whenever there are no messages left on the subscription.

**Comment by DXist** [comment]
> Just leaving my use case over here: a deployment of which the pods pull messages from gcloud pub/sub, autoscales on number of outstanding messages on the subscription, and should scale back to 0 pods whenever there are no messages left on the subscription.

I have similar use case and successfully implemented this behavior on premises. At the moment I'm trying to pass [review](https://github.com/kubernetes/kubernetes/pull/74526#issuecomment-499058825) of SIG API Machinery team.

**Comment by reyou** [comment]
Hey this is great @DXist. Just wondering, does it mean this feature will be available on the next release of k8s? So are we going to able to set deployment scalability between 0-N, and pods will not use any CPU or memory till they are awaken?

**Comment by DXist** [comment]
@reyou , scale-to-zero feature will be available if HPAScaleToZero feature gate is activated. The feature will be available for custom object or external metrics and not for resource metrics like CPU or memory.

**Comment by reyou** [comment]
Ah OK thanks for the clarification. I was thinking to have serverless behavior from the pods.
Seems like this is not the cause currently. Anyways great feature addon. Thanks!

**Comment by Enelar** [comment]
@DXist, I think @k8s-ci-bot misleading current status of the issue, which could lead to newly opened issues. My first impression was "it won't be released", and I was sad, read whole thread trying to understand "why".

In my case, I would like to start / pull messages from RabbitMQ queue, so there is no reason for preventing scale-to-zero. When first is ready, it pulls. If queue is empty long enough - it's dies. Current kubernetes is more than capable of providing that without any alterations. The single reason im unavailable to achieve desired functionality: validation check on HPA.

- Desired RabbitMQ queue length = 0
- HPA 0-100
- Done

**Comment by DXist** [comment]
@Enelar , my PR handled the validation of zero value of `minReplicas` for object/external metrics so you can wait for the next K8S release or prepare your own K8S build from current master.

**Comment by Enelar** [comment]
Whoops, sorry, made wrong assumptions, and did not read the code.
Great job, thanks!

**Comment by fpetkovski** [comment]
@DXist thank you for this contribution, it is provides enormous value!

**Comment by numbsafari** [comment]
Is there an issue tracking for getting this out of Alpha? I'm using GKE and would like to use this capability for pubsub/mq-based workloads as well, but cannot do so since this is gated with a feature flag? 

I would love to track whatever issues are blocking this from going GA.

**Comment by DXist** [comment]
@numbsafari , I've asked on [the PR page](https://github.com/kubernetes/kubernetes/pull/74526)

**Comment by memorais** [comment]
There are any tutorial or sample on how to configure this from adapter configuration to generate custom metrics or external metrics to configure the metric inside HPA?

**Comment by DXist** [comment]
@memorais , there is [the walkthrough](https://github.com/DirectXMan12/k8s-prometheus-adapter/blob/master/docs/walkthrough.md) for Prometheus Adapter.

**Comment by akrock** [comment]
Sure would be nice on some kind of guidance on what it takes to push `FeatureGates` through the pipeline.

Seems all of the major managed K8S  providers (AKS/EKS/GKE) have taken a similar stance on only allowing GA `FeatureGates`

**Comment by jeffreybrowning** [comment]
Haven't seen a wrap-up on this issue, excuse me if I missed something.

Is this still only in `Alpha`?

**Comment by DXist** [comment]
Technically the feature gate HPAScaleToZero was required to support backwards compatible cluster rollback to previous minor K8S version. There were several K8S releases after the feature was merged and in my opinion nothing stops to make the feature GA.

**Comment by numbsafari** [comment]
@gjtempleton @mwielgus - Looks like the work here is done, but is being blocked by a (no longer necessary) alpha feature gate. For those of us trying to implement worker-queue type workloads, this would be a massive benefit to be able to enable this in actual, working clusters.

Is there anything that can be done to get support in removing the alpha label here? Is it just a PR, or does some other process need to occur?

**Comment by gjtempleton** [comment]
Hey @numbsafari, I see from the discussion on the original PR that the next steps I'd suggest if you want to get this feature promoted have already been mentioned. 

The first step I'd recommend is someone raising this at the [SIG Autoscaling](https://docs.google.com/document/d/1RvhQAEIrVLHbyNnuaT99-6u9ZUMp7BfkPupT2LAZK7w) weekly meeting (ideally someone who's wanting to drive forward or be involved in the graduation of the feature.)

**Comment by jeffreybrowning** [comment]
@numbsafari @gjtempleton sorry for not reporting back here, I believe we have some movement here: https://github.com/kubernetes/kubernetes/pull/74526

Someone needs to write an upgrade document for alpha -> beta for this release. No one has stepped up yet.

**Comment by jiraguha** [comment]
Sorry, it is quite hard to follow, why this issue is closed. A solution is still ongoing?

**Comment by gjtempleton** [comment]
The solution exists, but currently only as an alpha feature, which needs to be explicitly enabled via a feature flag passed to a cluster.

As mentioned above, most managed service providers have a policy of only allowing GA features, so it's not available in a number of managed services currently. As mentioned in the discussions on the original PR to make it alpha mentioned above, it'll need to go through the [Kubernetes Enhancement Proposal](https://github.com/kubernetes/enhancements/tree/master/keps) (KEP) process before graduating from Alpha.

**Comment by johanneswuerbach** [comment]
I started the official propose here https://github.com/kubernetes/enhancements/issues/2021. Looking forward to any of your comments or feedback on the KEP itself here https://github.com/kubernetes/enhancements/pull/2022.

**Comment by vinayagg** [comment]
@johanneswuerbach can I suggest following use case which can enable scaling web apps to 0 pods.
Most, if not all, can be done with your proposal. A new ingress controller and a sliding scale counter needs to be implemented (or prometheus?). But I wanted to suggest this use case so you can review if your proposal/code needs some tweaking.

- Let's say there are 100 users, user1 to user100, each one has a deployment, WordpressDeployment1 to WordpressDeployment100.

- Each deploymet will scale based on a sliding scale of number of requests in last Y second. So somewhere there is a hashmap of "No Of Requests For Deployment X In Last X Seconds".

- Each Deployment's HPA will check, if noOfRequestsForDeploymentXInLastYSeconds > 0, it will scale to at least 1. Rest of scalability can be defined in HPA.

- On every request, Ingress controller will check, if number of pods for deploymentX > 0, it will forward request as intended to the right service.

- If number of pods for deploymentX < 1, then ingress controller will do following 
  -- notify the deployment to scale up, so that there is no delay in scaling up deployment
  -- redirect request to a different deployment. This deployment can have a page that says "redirecting, please wait ..." and does client side refresh every 1 seconds.

**Comment by gyliu513** [comment]
A really long discussion here :), does this feature has been delivered in Kubernetes? Any document that I can refer to? Thanks!

**Comment by DXist** [comment]
@gyliu513 , at the moment it's available only when

* HPAScaleToZero feature gate is enabled and 
* external or objects metrics are used to describe the scaled resource (like length of an external task queue, task ingress/egress ratio, etc.)

**Comment by Jeffwan** [comment]
I am trying to understand the difference between `HPAScaleToZero` and `KEDA`. KEDA supports 0<->1 as well. Besides event source, any difference?

**Comment by DXist** [comment]
> I am trying to understand the difference between `HPAScaleToZero` and `KEDA`. KEDA supports 0<->1 as well. Besides event source, any difference?

It looks like KEDA provides opinionated solution to scale workloads that run on top of different messaging services.

I've used HPAScaleToZero together with RabbitMQ. I configured two custom metrics.

One of them allowed to maintain N messages per worker after median filtering (N - is configured concurrency level per worker).
The second estimated scaling ratio from ingress and egress rates to scale rapidly in case of workload bursts and resource overloads.

Final setup was on top of bare metal and cloud resources and used consumer and message priorities.

**Comment by joebowbeer** [comment]
KEDA is built on top of the HPA and allows the user to leverage ExternalMetrics to define autoscaling criteria based on information from any event source.

Normally the HPA would do the scaling based on these internal and external event sources, but for the 0<->1 case the KEDA docs describe a custom ScaledObject resource, and it is KEDA that creates the HPA when the ScaledObject is created, and destroys the HPA when the ScaledObject is destroyed.

https://cloudblogs.microsoft.com/opensource/2020/05/12/scaling-kubernetes-keda-intro-kubernetes-based-event-driven-autoscaling/

**Comment by jiraguha** [comment]
@DXist for

> I've used HPAScaleToZero together with RabbitMQ. I configured two custom metrics.

Interesting, have you any resources/docs showing your setup?

**Comment by johanneswuerbach** [comment]
I had a talk at a k8s meetup on this https://github.com/johanneswuerbach/k8s-meetup-talk-autoscaling/, which has a documented setup.

It might be a bit outdated, but should give you an idea.

**Comment by DXist** [comment]
[Walkthrough](https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/docs/walkthrough.md) of setting up Prometheus Adapter is also relevant.

**Comment by machine424** [comment]
Hello,

I got tired of having nodes that are only used 3 hours a day and I implemented a little tool https://github.com/machine424/kube-hpa-scale-to-zero to simulate HPAScaleToZero especially managed Kubernetes clusters.

Awaiting for the feature gate to get promoted.

**Comment by ashak** [comment]
Is there anything people can do with to help this be promoted beyond alpha?

Not having this available in managed Kubernetes clusters has meant that we stopped using it in unmanaged Kubernetes clusters in favour of something that was consistent across all clusters. Ideally we'd like to go back to using it rather than a custom rolled solution similar to machine424's kube-hpa-scale-to-zero mentioned above, but until we can use it in our managed clusters that's unlikely to happen.

**Comment by DXist** [comment]
@ashak , if your use case is based on custom metrics you could step up and try to join a meeting of SIG autoscaling or SIG API machinery.

My current project doesn't require HPAScaleToZero feature gate - so a motivated person is needed to move it through the process.

**Comment by johanneswuerbach** [comment]
I've updated the linked KEP again, but overall its a fairly slow process. Let me know if there is anything on my side that can be done to speed this up.

Once the KEP is merged I'm happy to work on an implementation.

**Comment by haroldship** [comment]
I would like to enable this on my dev cluster (Rancher Desktop on Mac). How do I do it?

**Comment by lukaszsurfer** [comment]
@haroldship try enabling feature gate https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/, namely `HPAScaleToZero`

**Comment by haroldship** [comment]
Hi @lukaszsurfer
This is exactly what I don’t know how to do. Can I do it with kubectl? What is the exact command or procedure to enable this feature gate?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
